File: Auto_Use/web/llm_provider/perplexity/service.rs
// Copyright 2026 Ashish Yadav — Auto-Use

//! Perplexity Agent API provider (Responses-style dialect).

use serde_json::{json, Map, Value};

use super::view::get_reasoning_params;
use crate::agent::main_driver::view::dumps;
use crate::llm_provider::{api_url, args_dict, extract_text_first, post_json, SCREENSHOT_MEDIA_TYPE};

pub struct PerplexityProvider {
    pub api_key: String,
    pub cli_agent: bool,
    /// Responses-style flat function tools; None only for mode="text".
    pub tools: Option<Vec<Value>>,
}

impl PerplexityProvider {
    pub fn send_request(
        &self,
        messages: &[Value],
        model: &str,
        annotated_screenshot_base64: Option<&str>,
    ) -> Result<Value, String> {
        // Separate system prompt (top-level `instructions`) from the
        // conversation, translating the NATIVE TRANSCRIPT into the Responses
        // dialect: flat function_call / function_call_output items.
        let mut instructions: Option<String> = None;
        let mut input_messages: Vec<Value> = Vec::new();

        for msg in messages {
            let role = msg.get("role").and_then(Value::as_str).unwrap_or("");
            match role {
                "system" => {
                    let content = msg.get("content").cloned().unwrap_or(json!(""));
                    instructions = Some(match &content {
                        Value::Array(items) => items
                            .iter()
                            .find(|it| it.get("type").and_then(Value::as_str) == Some("text"))
                            .and_then(|it| it.get("text").and_then(Value::as_str))
                            .unwrap_or("")
                            .to_string(),
                        Value::String(s) => s.clone(),
                        other => crate::agent::main_driver::view::py_str_of(other),
                    });
                }
                "tool" => {
                    input_messages.push(json!({
                        "type": "function_call_output",
                        "call_id": msg.get("tool_call_id").and_then(Value::as_str).unwrap_or(""),
                        "output": extract_text_first(msg.get("content").unwrap_or(&Value::Null)),
                    }));
                }
                "assistant" => {
                    let text = extract_text_first(msg.get("content").unwrap_or(&Value::Null));
                    if !text.trim().is_empty() {
                        input_messages.push(json!({
                            "role": "assistant",
                            "content": [{"type": "output_text", "text": text}],
                        }));
                    }
                    for tc in msg
                        .get("tool_calls")
                        .and_then(Value::as_array)
                        .cloned()
                        .unwrap_or_default()
                    {
                        let fn_ = tc.get("function").cloned().unwrap_or(json!({}));
                        let raw = fn_.get("arguments").cloned().unwrap_or(Value::Null);
                        let arguments = match raw {
                            Value::String(s) => s,
                            Value::Null => dumps(&json!({})),
                            other => dumps(&other),
                        };
                        input_messages.push(json!({
                            "type": "function_call",
                            "call_id": tc.get("id").and_then(Value::as_str).unwrap_or(""),
                            "name": fn_.get("name").and_then(Value::as_str).unwrap_or(""),
                            "arguments": arguments,
                        }));
                    }
                }
                "user" => {
                    input_messages.push(json!({
                        "role": "user",
                        "content": [{"type": "input_text",
                                     "text": extract_text_first(msg.get("content").unwrap_or(&Value::Null))}],
                    }));
                }
                _ => {}
            }
        }

        // Screenshot on the last user message — `last` may be a flat
        // function_call* item with no "role" key.
        if let Some(shot) = annotated_screenshot_base64 {
            if !self.cli_agent && !input_messages.is_empty() {
                let last = input_messages.len() - 1;
                if input_messages[last].get("role").and_then(Value::as_str) == Some("user") {
                    if let Some(Value::Array(content)) = input_messages[last].get_mut("content") {
                        content.push(json!({
                            "type": "input_image",
                            "image_url": format!("data:{SCREENSHOT_MEDIA_TYPE};base64,{shot}"),
                        }));
                    }
                }
            }
        }

        let mut data = Map::new();
        data.insert("model".into(), json!(model));
        data.insert("input".into(), Value::Array(input_messages));
        data.insert("max_output_tokens".into(), json!(10000));
        if let Value::Object(extra) = get_reasoning_params(model) {
            for (k, v) in extra {
                data.insert(k, v);
            }
        }
        if let Some(instructions) = &instructions {
            if !instructions.is_empty() {
                data.insert("instructions".into(), json!(instructions));
            }
        }
        if let Some(tools) = &self.tools {
            data.insert("tools".into(), Value::Array(tools.clone()));
            // Canonical rationale in openrouter/service.rs.
            data.insert("tool_choice".into(), json!("required"));
        }

        let url = api_url("https://api.perplexity.ai/v1/agent");
        let auth = format!("Bearer {}", self.api_key);
        let headers = [
            ("Authorization", auth.as_str()),
            ("Content-Type", "application/json"),
        ];
        let resp = post_json(&url, &headers, &Value::Object(data))
            .map_err(|e| format!("Perplexity API request failed: {e}"))?;
        if resp.status >= 400 {
            return Err(format!(
                "Perplexity API request failed: HTTP {}\nResponse Body: {}",
                resp.status, resp.body
            ));
        }
        let result: Value = serde_json::from_str(&resp.body)
            .map_err(|e| format!("Perplexity API request failed: {e}"))?;

        // Normalize to choices[0].message.content format.
        let mut text_content = String::new();
        let mut calls: Vec<Value> = Vec::new();
        for output_item in result.get("output").and_then(Value::as_array).cloned().unwrap_or_default() {
            match output_item.get("type").and_then(Value::as_str) {
                Some("message") => {
                    for content_block in output_item
                        .get("content")
                        .and_then(Value::as_array)
                        .cloned()
                        .unwrap_or_default()
                    {
                        if content_block.get("type").and_then(Value::as_str)
                            == Some("output_text")
                            && text_content.is_empty()
                        {
                            text_content = content_block
                                .get("text")
                                .and_then(Value::as_str)
                                .unwrap_or("")
                                .to_string();
                        }
                    }
                }
                Some("function_call") if self.tools.is_some() => {
                    let args = args_dict(output_item.get("arguments"));
                    let id = ["call_id", "id"]
                        .iter()
                        .find_map(|k| {
                            let v = output_item.get(*k)?;
                            let s = crate::agent::main_driver::view::py_str_of(v);
                            if s.is_empty() || s == "None" { None } else { Some(s) }
                        })
                        .unwrap_or_else(|| format!("call_{}", calls.len()));
                    calls.push(json!({
                        "id": id,
                        "name": output_item.get("name").and_then(Value::as_str).unwrap_or(""),
                        "arguments": args,
                    }));
                }
                _ => {}
            }
        }

        let mut message = Map::new();
        message.insert("content".into(), json!(text_content));
        if self.tools.is_some() {
            message.insert("tool_calls".into(), Value::Array(calls));
        }
        Ok(json!({
            "choices": [{"message": Value::Object(message)}],
            "usage": result.get("usage").cloned().unwrap_or(json!({})),
        }))
    }
}

File: Auto_Use/ios/llm_provider/perplexity/service.py
# Copyright 2026 Cursortouch — Auto-Use

import json
import requests
from typing import Dict, Any, Optional

from .view import get_model_info, get_reasoning_params
from .. import LLM_HTTP_TIMEOUT


class PerplexityProvider:
    """Perplexity Agent API provider for LLM interactions"""

    def __init__(self, api_key: str, cli_agent: bool = False, model_info: dict = None, tools: list = None):
        self.api_key = api_key
        self.api_url = "https://api.perplexity.ai/v1/agent"
        self.cli_agent = cli_agent
        self.model_info = model_info or {}
        # Native tool calling: Responses-style flat function tools ({type,
        # name, description, parameters}) - the model's output contract.
        # None only for mode="text" (plain prose).
        self.tools = tools or None

    def send_request(self, messages: list, model: str, annotated_screenshot_base64: Optional[str] = None) -> Dict[str, Any]:
        """Send request to Perplexity Agent API"""
        
        # Separate system prompt from conversation messages
        instructions = None
        input_messages = []
        
        for msg in messages:
            if msg["role"] == "system":
                content = msg["content"]
                if isinstance(content, list):
                    for item in content:
                        if isinstance(item, dict) and item.get("type") == "text":
                            instructions = item.get("text", "")
                            break
                else:
                    instructions = content
            elif msg["role"] == "tool":
                # NATIVE TRANSCRIPT: the Responses dialect carries a tool
                # result as its own flat output item keyed by call_id.
                input_messages.append({
                    "type": "function_call_output",
                    "call_id": msg.get("tool_call_id") or "",
                    "output": self._extract_text(msg.get("content")),
                })
            elif msg["role"] == "assistant":
                text = self._extract_text(msg["content"])
                if text.strip():
                    input_messages.append({"role": "assistant", "content": [{"type": "output_text", "text": text}]})
                # ...and one function_call item per tool call the model made,
                # so it sees its OWN calls replayed in the canonical shape.
                for tc in (msg.get("tool_calls") or []):
                    fn = (tc or {}).get("function") or {}
                    raw = fn.get("arguments")
                    input_messages.append({
                        "type": "function_call",
                        "call_id": (tc or {}).get("id") or "",
                        "name": fn.get("name") or "",
                        "arguments": raw if isinstance(raw, str) else json.dumps(raw or {}),
                    })
            elif msg["role"] == "user":
                text = self._extract_text(msg["content"])
                input_messages.append({"role": "user", "content": [{"type": "input_text", "text": text}]})

        # Add screenshot to last user message if provided and NOT cli_agent
        if annotated_screenshot_base64 and not self.cli_agent and len(input_messages) > 0:
            last = input_messages[-1]
            # `last` may be a flat function_call* item with no "role" key.
            if last.get("role") == "user":
                last["content"].append({
                    "type": "input_image",
                    "image_url": f"data:image/jpeg;base64,{annotated_screenshot_base64}"
                })
        
        headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json"
        }
        
        data = {
            "model": model,
            "input": input_messages,
            "max_output_tokens": 10000,
        }

        # How hard the model thinks, per view.py's per-model table, carried as
        # the Agent API's reasoning config. Omitted for a hand-typed model
        # name, which then keeps Perplexity's own default.
        data.update(get_reasoning_params(model))
        
        if instructions:
            data["instructions"] = instructions

        # No tools (mode="text") -> plain text: omit the tool params entirely.
        if self.tools:
            data["tools"] = self.tools
            # "required" - every turn must call at least one tool; a text-only
            # turn is never a valid step (even termination is the `done`
            # tool). Responses-dialect value; if Perplexity's validator ever
            # rejects it, the request 400s loudly - flip back to "auto".
            # Canonical rationale: see openrouter/service.py.
            data["tool_choice"] = "required"

        try:
            response = requests.post(self.api_url, json=data, headers=headers, timeout=LLM_HTTP_TIMEOUT)
            response.raise_for_status()
            result = response.json()

            # Normalize to choices[0].message.content format
            text_content = ""
            calls = []
            for output_item in result.get("output", []):
                if output_item.get("type") == "message":
                    for content_block in output_item.get("content", []):
                        if content_block.get("type") == "output_text" and not text_content:
                            text_content = content_block.get("text", "")
                elif self.tools and output_item.get("type") == "function_call":
                    # Responses dialect: arguments is a JSON string
                    # (malformed -> {}).
                    raw_args = output_item.get("arguments")
                    if isinstance(raw_args, dict):
                        args = raw_args
                    else:
                        try:
                            args = json.loads(raw_args or "{}")
                        except Exception:
                            args = {}
                    if not isinstance(args, dict):
                        args = {}
                    calls.append({
                        "id": str(output_item.get("call_id") or output_item.get("id") or "")
                              or f"call_{len(calls)}",
                        "name": output_item.get("name") or "",
                        "arguments": args,
                    })

            message: Dict[str, Any] = {"content": text_content}
            if self.tools:
                message["tool_calls"] = calls
            return {
                "choices": [{
                    "message": message
                }],
                "usage": result.get("usage", {}),
            }
            
        except requests.exceptions.RequestException as e:
            error_msg = f"Perplexity API request failed: {str(e)}"
            if hasattr(e, 'response') and e.response is not None:
                error_msg += f"\nResponse Body: {e.response.text}"
            raise Exception(error_msg)
    
    @staticmethod
    def _extract_text(content) -> str:
        """Extract text from message content (string or list format with cache_control)"""
        if isinstance(content, str):
            return content
        if isinstance(content, list):
            for item in content:
                if isinstance(item, dict) and item.get("type") == "text":
                    return item.get("text", "")
            return str(content)
        return str(content)